### PR TITLE
5164 hook support

### LINKED ISSUE #2417: Create a EIP5164 Hook and ISM
There are existing bridges that have EIP5164 adapters, you can find some in https://github.com/MultiMessageAggregation/multibridge/tree/main/contracts/adapters

This ticket captures the creation of a 5164 hook and ISM that allows Hyperlane users to leverage those implementations for use, ideally in an aggregation ISM setup as desired by the Uniswap folks.

This allows us to collect feedback on which bridges folks want to use, and then more native hooks and ISMs can be provided with that feedback.

### PR DESCRIPTION
**Pull request by aroralanuk**
### Description

- Adding support for any ERC-5164 compatible bridge adapter to be added as a hook. See https://eips.ethereum.org/EIPS/eip-5164 for more details on the EIP and https://github.com/MultiMessageAggregation/multibridge for adapter implementations for different vendors.

### Drive-by changes

- Switched mapping(bytes32 => mapping(address => true) to mapping(bytes32 => bytes) for verifedMessageIds because the last boolean is redundant in the AbtrsactNativeISM contract.

### Related issues

- Fixes #2417 

### Backward compatibility

_Are these changes backward compatible?_

Yes

_Are there any infrastructure implications, e.g. changes that would prohibit deploying older commits using this infra tooling?_

None

### Testing

_What kind of testing have these changes undergone?_

Unit Tests

### CODE REVIEW

File: solidity/test/isms/erc5164/ERC5164ISM.t.sol [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+            address(dispatcher),
+            address(ism)
+        );
+    }
+
+    ///////////////////////////////////////////////////////////////////
+    ///                            TESTS                            ///
+    ///////////////////////////////////////////////////////////////////
+
+    function test_RevertWhen_Constructor() public {
```
Comment (yorhodes) [inline solidity/test/isms/erc5164/ERC5164ISM.t.sol:64]:
```suggestion
    function test_Constructor() public {
```

File: solidity/test/isms/erc5164/ERC5164ISM.t.sol [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+        vm.expectRevert("ERC5164Hook: invalid ISM");
+        hook = new ERC5164MessageHook(
+            TEST2_DOMAIN,
+            address(dispatcher),
+            address(0)
+        );
+    }
+
+    function test_PostDispatch() public {
```
[the remark this answers is not in the document]
Comment (yorhodes) [reply solidity/test/isms/erc5164/ERC5164ISM.t.sol:82]: here and everywhere

File: solidity/test/isms/erc5164/MockERC5164.sol [hunk 3]
```diff
@@ -0,0 +1,26 @@
+// SPDX-License-Identifier: MIT or Apache-2.0
+pragma solidity ^0.8.13;
+
+import {IMessageDispatcher} from "../../../contracts/hooks/ERC5164/interfaces/IMessageDispatcher.sol";
+
+contract MockMessageDispatcher is IMessageDispatcher {
```
Comment (yorhodes) [inline solidity/test/isms/erc5164/MockERC5164.sol:6]: nice
I think we have a `mock` directory

File: solidity/test/isms/erc5164/MockERC5164.sol [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        bytes32 messageId = keccak256(abi.encodePacked(toChainId, to, data));
+
+        // simulate a successful dispatch
+        emit MessageDispatched(messageId, msg.sender, toChainId, to, data);
+
+        return messageId;
+    }
+}
+
+contract MockMessageExecutor {
```
Comment (yorhodes) [inline solidity/test/isms/erc5164/MockERC5164.sol:21]: what is this?

File: solidity/test/isms/erc5164/MockMessageDispatcher.sol [hunk 5]
```diff
@@ -0,0 +1,19 @@
+// SPDX-License-Identifier: MIT or Apache-2.0
+pragma solidity ^0.8.13;
+
+import {IMessageDispatcher} from "../../../contracts/hooks/ERC5164/interfaces/IMessageDispatcher.sol";
+
+contract MockMessageDispatcher is IMessageDispatcher {
```
Comment (yorhodes) [inline solidity/test/isms/erc5164/MockMessageDispatcher.sol:6]: you duplicated this
move to mocks dir

**Review by yorhodes — APPROVED** [review]
please address my comments before merging

File: solidity/test/isms/erc5164/ERC5164ISM.t.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/test/isms/erc5164/ERC5164ISM.t.sol:82]: okk, I was following msolomon's foundry test convention

**Pull request merged**